junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestMethodTestDescriptor.java

```diff
@@ -231,4 +237,53 @@ private void invokeAfterEachCallbacks(JupiterEngineExecutionContext context) {
 		});
 	}
 
+	@Override
+	public void nodeSkipped(JupiterEngineExecutionContext context, TestDescriptor descriptor, SkipResult result) {
+		if (context != null) {
+			invokeTestWatchers(context, watcher -> {
+				watcher.testDisabled(context.getExtensionContext(), result.getReason());
+			}, false);
+		}
+	}
+
+	@Override
+	public void nodeFinished(JupiterEngineExecutionContext context, TestDescriptor descriptor,
+			TestExecutionResult result) {
+		if (context != null) {
+			invokeTestWatchers(context, watcher -> {
+				ExtensionContext extensionContext = context.getExtensionContext();
+				TestExecutionResult.Status status = result.getStatus();
+				switch (status) {
+					case SUCCESSFUL:
+						watcher.testSuccessful(extensionContext);
+						break;
+					case FAILED:
+						watcher.testFailed(extensionContext, result.getThrowable().get());
+						break;
+					case ABORTED:
+						watcher.testAborted(extensionContext, result.getThrowable().get());
+						break;
+				}
+			}, true);
+		}
+	}
+
+	private void invokeTestWatchers(JupiterEngineExecutionContext context, Consumer<TestWatcher> callback,
+			boolean reverseOrder) {
+		ExtensionRegistry registry = context.getExtensionRegistry();
+		List<? extends TestWatcher> watchers = reverseOrder ? registry.getReversedExtensions(TestWatcher.class)
+				: registry.getExtensions(TestWatcher.class);
+		watchers.forEach(watcher -> {
+			try {
+				callback.accept(watcher);
+			}
+			catch (Exception e) {
+				logger.warn(() -> {
+					return String.format("Error invoking TestWatcher %s for test %s : %s", watcher.getClass().getName(),
```

Comment:
This is broken. There are 3 occurrences of `%s`. Please pass the exception as an argument to the `warn()` method, and include the "test" in the message.

Reply:
Also, please introduce tests that verify the expected exception handling.

Reply:
Furthermore, we should likely catch `Throwable`, rethrow it if it's _blacklisted_, and otherwise swallow it (but log it, as it is now).